File: agent/tools/MCP/mcp-1panel/src/mcp_1panel/tools/websites.py
"""MCP tools for website, SSL, and DNS management via 1Panel API."""

from typing import Any, Callable, Dict, List
from mcp_1panel.tools.helpers import (
    icon_green,
    icon_red,
    icon_lock,
    icon_unlock,
    icon_status,
    header,
    fmt_val,
    fmt_obj,
    fmt_list,
    fmt_search,
    fmt_generic,
)


def register_tools(mcp, get_client, handlers=None):
    """Register all website/SSL/DNS management MCP tools."""

    # ------------------------------------------------------------------ #
    # Website list and detail
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_websites(page: int = 1, size: int = 20) -> str:
        """网站列表 — 域名、别名、运行状态、SSL"""
        p = get_client()
        data = p.post("/websites/search", {
            "page": page, "pageSize": size,
            "orderBy": "createdAt", "order": "descending",
        })
        lines, items = fmt_search(data, "网站列表")
        for w in items:
            domain = w.get("primaryDomain", w.get("domain", "?"))
            alias = w.get("alias", "")
            running = w.get("status", w.get("running", 0))
            run_icon = icon_green() if running else "\u23f9"  # 🟢 / ⏹
            https = w.get("https", w.get("hasSSL", False))
            https_icon = icon_lock() if https else icon_unlock()
            label = f"{run_icon} {https_icon} {domain}"
            if alias:
                label += f"  ({alias})"
            lines.append(f"  {label}")
        if not items:
            lines.append("  (空)")
        return "\n".join(lines)

    @mcp.tool()
    def panel_website_detail(id: int) -> str:
        """网站详情 — 配置、状态、域名"""
        p = get_client()
        data = p.get(f"/websites/detail?id={id}")
        return fmt_generic(data, f"网站详情 (ID: {id})")

    @mcp.tool()
    def panel_website_nginx(id: int, type: str = "nginx") -> str:
        """网站 Nginx 配置"""
        p = get_client()
        data = p.get(f"/websites/nginx?id={id}&type={type}")
        return fmt_generic(data, f"Nginx 配置 (ID: {id})")

    @mcp.tool()
    def panel_website_https(id: int) -> str:
        """网站 HTTPS 配置"""
        p = get_client()
        data = p.get(f"/websites/https?id={id}")
        return fmt_generic(data, f"HTTPS 配置 (ID: {id})")

    @mcp.tool()
    def panel_website_domains(id: int) -> str:
        """网站域名列表"""
        p = get_client()
        data = p.get(f"/websites/domains?id={id}")
        lines = [header(f"域名列表 (ID: {id})")]
        if isinstance(data, list):
            for d in data:
                domain = d.get("domain", d.get("name", fmt_val(d)))
                port = d.get("port", "")
                lines.append(f"  {domain}  {f':{port}' if port else ''}")
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    @mcp.tool()
    def panel_website_ssl(id: int) -> str:
        """网站 SSL 证书"""
        p = get_client()
        data = p.get(f"/websites/ssl?id={id}")
        return fmt_generic(data, f"SSL 证书 (ID: {id})")

    @mcp.tool()
    def panel_website_ssl_by_site(id: int) -> str:
        """按网站 ID 查询 SSL 证书"""
        p = get_client()
        data = p.get(f"/websites/ssl/bysite?id={id}")
        return fmt_generic(data, f"站点 SSL 证书 (ID: {id})")

    # ------------------------------------------------------------------ #
    # SSL, ACME, DNS, CA lists
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_ssl_list() -> str:
        """SSL 证书列表"""
        p = get_client()
        data = p.get("/websites/ssl/list")
        lines = [header("SSL 证书列表", len(data) if isinstance(data, list) else None)]
        if isinstance(data, list):
            for cert in data:
                domain = cert.get("primaryDomain", cert.get("domain", "?"))
                issuer = cert.get("issuer", cert.get("organization", ""))
                expire = cert.get("expireDate", cert.get("expireAt", "?"))
                lines.append(f"  {domain}  {fmt_val(issuer)}  expires: {expire}")
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    @mcp.tool()
    def panel_acme_accounts() -> str:
        """ACME 账户列表 — Let's Encrypt 等"""
        p = get_client()
        data = p.get("/websites/acme/accounts")
        lines = [header("ACME 账户", len(data) if isinstance(data, list) else None)]
        if isinstance(data, list):
            for acc in data:
                email = acc.get("email", "?")
                provider = acc.get("ca", acc.get("provider", "?"))
                status = acc.get("status", "?")
                st = icon_green() if status in ("valid", "active", "ok") else icon_red()
                lines.append(f"  {st} {email}  [{provider}]")
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    @mcp.tool()
    def panel_dns_accounts() -> str:
        """DNS 账户列表 — 域名解析服务商"""
        p = get_client()
        data = p.get("/websites/dns/accounts")
        lines = [header("DNS 账户", len(data) if isinstance(data, list) else None)]
        if isinstance(data, list):
            for acc in data:
                name = acc.get("name", acc.get("provider", "?"))
                vendor = acc.get("type", acc.get("vendor", "?"))
                status = acc.get("status", "")
                st = icon_green() if status in ("valid", "active", "ok") else icon_red() if status else ""
                label = f"  {name}  [{vendor}]"
                if st:
                    label = f"  {st} {name}  [{vendor}]"
                lines.append(label)
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    @mcp.tool()
    def panel_ca_certificates() -> str:
        """CA 证书列表 — 自定义证书颁发机构"""
        p = get_client()
        data = p.get("/websites/ca")
        lines = [header("CA 证书", len(data) if isinstance(data, list) else None)]
        if isinstance(data, list):
            for ca in data:
                name = ca.get("name", ca.get("commonName", "?"))
                org = ca.get("organization", "")
                expire = ca.get("expireDate", ca.get("expireAt", ""))
                ln = f"  {name}"
                if org:
                    ln += f"  [{org}]"
                if expire:
                    ln += f"  expires: {expire}"
                lines.append(ln)
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    @mcp.tool()
    def panel_ca_detail(id: int) -> str:
        """CA 证书详情"""
        p = get_client()
        data = p.get(f"/websites/ca/detail?id={id}")
        return fmt_generic(data, f"CA 证书详情 (ID: {id})")

    # ------------------------------------------------------------------ #
    # Database and config
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_website_db() -> str:
        """网站关联数据库"""
        p = get_client()
        data = p.get("/websites/db")
        lines = [header("网站关联数据库", len(data) if isinstance(data, list) else None)]
        if isinstance(data, list):
            for db in data:
                name = db.get("name", db.get("database", "?"))
                wname = db.get("website", db.get("siteName", ""))
                lines.append(f"  {name}  {f'({wname})' if wname else ''}")
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    @mcp.tool()
    def panel_website_config(id: int) -> str:
        """网站 Nginx 配置详情"""
        p = get_client()
        data = p.post("/websites/config", {"id": id})
        return fmt_generic(data, f"Nginx 配置详情 (ID: {id})")

    # ------------------------------------------------------------------ #
    # Directory, CORS, RealIP
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_website_dir(id: int) -> str:
        """网站目录信息"""
        p = get_client()
        data = p.get(f"/websites/dir?id={id}")
        return fmt_generic(data, f"网站目录 (ID: {id})")

    @mcp.tool()
    def panel_website_cors(id: int) -> str:
        """网站 CORS 配置"""
        p = get_client()
        data = p.get(f"/websites/cors?id={id}")
        return fmt_generic(data, f"CORS 配置 (ID: {id})")

    @mcp.tool()
    def panel_website_realip(id: int) -> str:
        """网站 RealIP 配置"""
        p = get_client()
        data = p.get(f"/websites/realip?id={id}")
        return fmt_generic(data, f"RealIP 配置 (ID: {id})")

    # ------------------------------------------------------------------ #
    # Upstreams, Proxy, Rewrite, Redirect
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_website_upstreams() -> str:
        """负载均衡上游列表"""
        p = get_client()
        data = p.get("/websites/upstreams")
        lines = [header("上游列表", len(data) if isinstance(data, list) else None)]
        if isinstance(data, list):
            for up in data:
                name = up.get("name", up.get("upstream", "?"))
                servers = up.get("servers", [])
                ln = f"  {name}"
                if isinstance(servers, list):
                    for s in servers:
                        addr = s.get("address", fmt_val(s))
                        ln += f"  -> {addr}"
                lines.append(ln)
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    @mcp.tool()
    def panel_website_proxy(id: int) -> str:
        """网站反向代理配置"""
        p = get_client()
        data = p.get(f"/websites/proxy?id={id}")
        return fmt_generic(data, f"反向代理 (ID: {id})")

    @mcp.tool()
    def panel_website_rewrite(id: int) -> str:
        """网站重写规则"""
        p = get_client()
        data = p.get(f"/websites/rewrite?id={id}")
        return fmt_generic(data, f"重写规则 (ID: {id})")

    @mcp.tool()
    def panel_website_redirect(id: int) -> str:
        """网站重定向配置"""
        p = get_client()
        data = p.get(f"/websites/redirect?id={id}")
        return fmt_generic(data, f"重定向 (ID: {id})")

    # ------------------------------------------------------------------ #
    # Anti-leech, Auth, Log
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_website_antileech(id: int) -> str:
        """网站防盗链配置"""
        p = get_client()
        data = p.get(f"/websites/antileech?id={id}")
        return fmt_generic(data, f"防盗链 (ID: {id})")

    @mcp.tool()
    def panel_website_auth(id: int) -> str:
        """网站 BasicAuth 认证配置"""
        p = get_client()
        data = p.get(f"/websites/auth?id={id}")
        return fmt_generic(data, f"BasicAuth (ID: {id})")

    @mcp.tool()
    def panel_website_log(id: int, rows: int = 50) -> str:
        """网站访问日志"""
        p = get_client()
        data = p.get(f"/websites/log?id={id}&rows={rows}")
        lines = [header(f"访问日志 (ID: {id})")]
        if isinstance(data, list):
            for log in data[:rows]:
                ts = log.get("time", log.get("createdAt", log.get("timestamp", "?")))
                ip = log.get("ip", log.get("remoteAddr", ""))
                method = log.get("method", log.get("requestMethod", ""))
                url = log.get("url", log.get("requestUri", ""))
                status = log.get("status", log.get("httpCode", ""))
                ln = f"  {ts}"
                if ip:
                    ln += f"  {ip}"
                if method:
                    ln += f"  {method}"
                if url:
                    ln += f"  {url}"
                if status:
                    ln += f"  [{status}]"
                lines.append(ln)
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    # ------------------------------------------------------------------ #
    # Simple list
    # ------------------------------------------------------------------ #

    @mcp.tool()
    def panel_website_list_simple() -> str:
        """网站简要列表"""
        p = get_client()
        data = p.get("/websites/list")
        lines = [header("网站简要列表", len(data) if isinstance(data, list) else None)]
        if isinstance(data, list):
            for w in data:
                domain = w.get("primaryDomain", w.get("domain", w.get("name", "?")))
                alias = w.get("alias", "")
                ln = f"  {domain}"
                if alias:
                    ln += f"  ({alias})"
                lines.append(ln)
        elif isinstance(data, dict):
            lines.extend(fmt_obj(data))
        else:
            lines.append(f"  {fmt_val(data)}")
        if not data:
            lines.append("  (空)")
        return "\n".join(lines)

    # -- 收集 handler 供 resources 复用 --
    if handlers is not None:
        for _name, _fn in list(locals().items()):
            if _name.startswith("panel_") and callable(_fn):
                handlers[_name] = _fn

File: agent/tools/MCP/npm/mcp-1panel/src/tools/websites.js
/**
 * mcp-1panel — Websites 模块工具（25 个）
 */
import { tool, toolWithParams, fmtObj, fmtList, fmtSearch, statusIcon } from "./helpers.js";
import { api } from "../api-proxy.js";

export const WEBSITES_TOOLS = {

  panel_websites: tool("网站列表 — 域名、SSL 状态、别名", async (args) => {
    const d = await api.post("/websites/search", {
      page: args.page ?? 1, pageSize: args.size ?? 20,
      orderBy: "createdAt", order: "descending",
    });
    const items = d.items ?? [];
    const lines = [`= 网站 (共${d.total ?? 0}个) =`];
    for (const w of items) {
      const s = w.status === "Running" ? "🟢" : "🔴";
      const h = w.https ? "🔒" : "🔓";
      lines.push(`  ${s} ${h} ${w.primaryDomain ?? "?"} [${w.alias ?? "?"}]`);
    }
    if (!items.length) lines.push("  (空)");
    return lines.join("\n");
  }, { page: { type: "number", default: 1 }, size: { type: "number", default: 20 } }),

  panel_website_detail: toolWithParams("网站详情 — 配置、状态、域名", async (args) => {
    const d = await api.get(`/websites/${args.id}`);
    return fmtObj("网站详情", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_nginx: toolWithParams("网站 Nginx 配置", async (args) => {
    const d = await api.get(`/websites/${args.id}/config/${args.type ?? "nginx"}`);
    return typeof d === "string" ? `= Nginx 配置 =\n${d.slice(0, 3000)}` : fmtObj("Nginx 配置", d, 200);
  }, { id: { type: "number", default: 1 }, type: { type: "string", default: "nginx" } }),

  panel_website_https: toolWithParams("网站 HTTPS 配置", async (args) => {
    const d = await api.get(`/websites/${args.id}/https`);
    return fmtObj("HTTPS 配置", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_domains: toolWithParams("网站域名列表", async (args) => {
    const d = await api.get(`/websites/domains/${args.id}`);
    return fmtList("网站域名", Array.isArray(d) ? d : [], (dm) => `${dm.domain ?? "?"}  ${dm.port ?? "?"}  ${dm.https ? "🔒" : "🔓"}`);
  }, { id: { type: "number", default: 1 } }),

  panel_website_ssl: toolWithParams("网站 SSL 证书", async (args) => {
    const d = await api.get(`/websites/ssl/${args.id}`);
    return fmtObj("SSL 证书", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_ssl_by_site: toolWithParams("按网站 ID 查询 SSL 证书", async (args) => {
    const d = await api.get(`/websites/ssl/website/${args.id}`);
    return fmtObj("网站 SSL", d);
  }, { id: { type: "number", default: 1 } }),

  panel_ssl_list: tool("SSL 证书列表", async () => {
    const d = await api.post("/websites/ssl/search", { page: 1, pageSize: 50 });
    return fmtSearch("SSL 证书", d, (s) => `${s.domains ?? "?"}  到期: ${s.expireDate ?? "?"}  颁发者: ${s.issuer ?? "?"}`);
  }),

  panel_acme_accounts: tool("ACME 账户列表 — Let's Encrypt 等", async () => {
    const d = await api.post("/websites/acme/search", { page: 1, pageSize: 50 });
    return fmtSearch("ACME 账户", d, (a) => `${a.email ?? "?"}  ${a.type ?? "?"}  ${a.status ?? "?"}`);
  }),

  panel_dns_accounts: tool("DNS 账户列表 — 域名解析服务商", async () => {
    const d = await api.post("/websites/dns/search", { page: 1, pageSize: 50 });
    return fmtSearch("DNS 账户", d, (dns) => `${dns.name ?? "?"}  ${dns.type ?? "?"}  ${dns.status ?? "?"}`);
  }),

  panel_ca_certificates: tool("CA 证书列表 — 自定义证书颁发机构", async () => {
    const d = await api.post("/websites/ca/search", { page: 1, pageSize: 50 });
    return fmtSearch("CA 证书", d, (ca) => `${ca.name ?? "?"}  ${ca.commonName ?? "?"}  到期: ${ca.expireDate ?? "?"}`);
  }),

  panel_ca_detail: toolWithParams("CA 证书详情", async (args) => {
    const d = await api.get(`/websites/ca/${args.id}`);
    return fmtObj("CA 证书", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_db: tool("网站关联数据库", async () => {
    const d = await api.get(`/websites/databases`);
    return fmtList("网站数据库", Array.isArray(d) ? d : [], (db) => `${db.name ?? "?"}  ${db.type ?? "?"}  ${db.status ?? "?"}`);
  }),

  panel_website_config: toolWithParams("网站 Nginx 配置详情", async (args) => {
    const d = await api.post("/websites/config", { websiteId: args.id });
    return fmtObj("网站 Nginx 配置", d, 200);
  }, { id: { type: "number", default: 1 } }),

  panel_website_dir: toolWithParams("网站目录信息", async (args) => {
    const d = await api.post("/websites/dir", { websiteId: args.id });
    return fmtObj("网站目录", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_cors: toolWithParams("网站 CORS 配置", async (args) => {
    const d = await api.get(`/websites/cors/${args.id}`);
    return fmtObj("CORS 配置", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_realip: toolWithParams("网站 RealIP 配置", async (args) => {
    const d = await api.get(`/websites/realip/config/${args.id}`);
    return fmtObj("RealIP 配置", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_upstreams: tool("负载均衡上游列表", async () => {
    const d = await api.get("/websites/lbs");
    return fmtList("负载均衡上游", Array.isArray(d) ? d : [], (lb) => `${lb.name ?? "?"}  ${lb.type ?? "?"}  ${lb.status ?? "?"}`);
  }),

  panel_website_proxy: toolWithParams("网站反向代理配置", async (args) => {
    const d = await api.post("/websites/proxies", { websiteId: args.id });
    return fmtObj("反向代理", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_rewrite: toolWithParams("网站重写规则", async (args) => {
    const d = await api.post("/websites/rewrite", { websiteId: args.id });
    return fmtObj("重写规则", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_redirect: toolWithParams("网站重定向配置", async (args) => {
    const d = await api.post("/websites/redirect", { websiteId: args.id });
    return fmtObj("重定向", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_antileech: toolWithParams("网站防盗链配置", async (args) => {
    const d = await api.post("/websites/leech", { websiteId: args.id });
    return fmtObj("防盗链", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_auth: toolWithParams("网站 BasicAuth 认证配置", async (args) => {
    const d = await api.post("/websites/auths", { websiteId: args.id });
    return fmtObj("BasicAuth", d);
  }, { id: { type: "number", default: 1 } }),

  panel_website_log: toolWithParams("网站访问日志", async (args) => {
    const d = await api.post("/websites/log", { websiteId: args.id, page: 1, pageSize: args.rows ?? 50 });
    return fmtObj("网站日志", d, 200);
  }, { id: { type: "number", default: 1 }, rows: { type: "number", default: 50 } }),

  panel_website_list_simple: tool("网站简要列表", async () => {
    const d = await api.get("/websites/list");
    return fmtList("网站列表", Array.isArray(d) ? d : [], (w) => `${w.domain ?? w.primaryDomain ?? "?"}  ${statusIcon(w.status)} ${w.status ?? "?"}`);
  }),

};
